mirgecom/profiling.py

```diff
@@ -0,0 +1,208 @@
+            # N.B.: The invoker code might be different for the same program with
+            # different kwargs
+            if (program not in self.kernel_stats
+              or args_tuple not in self.kernel_stats[program]):
+                executor = program.target.get_kernel_executor(program, self.queue)
+                info = executor.kernel_info(executor.arg_to_dtype_set(kwargs))
+
+                kernel = executor.get_typed_and_scheduled_kernel(
+                    executor.arg_to_dtype_set(kwargs))
+
+                data = info.implemented_data_info
+
+                # Generate the wrapper code
+                wrapper = executor.get_wrapper_generator()
+
+                gen = PythonCodeGenerator()
+
+                wrapper.generate_integer_arg_finding_from_shapes(gen, kernel, data)
+                wrapper.generate_integer_arg_finding_from_offsets(gen, kernel, data)
+                wrapper.generate_integer_arg_finding_from_strides(gen, kernel, data)
+
+                types = {}
+                param_dict = {}
+
+                for key, value in kwargs.items():
+                    if hasattr(value, 'dtype') and not value.dtype == object:
+                        types[key] = value.dtype
+                    param_dict[key] = value
+
+                for key, value in kernel.arg_dict.items():
+                    if key not in param_dict:
+                        param_dict[key] = None
+
+                for d in data:
+                    if d.name not in param_dict:
+                        param_dict[d.name] = None
+
+                # Run the wrapper code, save argument values in param_dict
+                exec(gen.get(), param_dict)
+
+                # Get flops/memory statistics
+                kernel.options.ignore_boostable_into = True
+                kernel = lp.add_and_infer_dtypes(kernel, types)
+                op_map = lp.get_op_map(kernel, subgroup_size='guess')
+                mem_map = lp.get_mem_access_map(kernel, subgroup_size='guess')
+
+                f32op = op_map.filter_by(dtype=[np.float32]).eval_and_sum(param_dict)
+                f64op = op_map.filter_by(dtype=[np.float64]).eval_and_sum(param_dict)
+
+                mem32 = mem_map.filter_by(dtype=[np.float32]).eval_and_sum(
+                    param_dict)
+                mem64 = mem_map.filter_by(dtype=[np.float64]).eval_and_sum(
+                    param_dict)
+
+                flops = f32op + f64op
+                bytes_accessed = mem32*4 + mem64*8
+
+                if program not in self.kernel_stats:
+                    self.kernel_stats[program] = {}
+
+                self.kernel_stats[program][args_tuple] = (flops, bytes_accessed)
+
+        evt, result = program(self.queue, **kwargs, allocator=self.allocator)
+
+        if self.profiling_enabled:
+            flops = self.kernel_stats[program][args_tuple][0]
+            bytes_accessed = self.kernel_stats[program][args_tuple][1]
```

Comment:
Why is this here? The variables are still set from above? Also, use tuple unpacking, not indexing.

Reply:
Variables were set above only when the `if` was true. Clarified the code a bit by putting this in an `else`.